game-core/src/main/java/games/strategy/triplea/util/TuvUtils.java

```diff
@@ -134,6 +133,15 @@ private TuvUtils() {}
       final int averagedCost = (int) Math.round(((double) totalCosts / (double) costsForType.size()));
       costs.put(ut, averagedCost);
     }
+
+    // Add any units that have XML TUV even if they aren't purchasable
+    for (final UnitType unitType : data.getUnitTypeList()) {
+      final UnitAttachment ua = UnitAttachment.get(unitType);
+      if (ua != null && ua.getTuv() > 0) {
+        costs.put(unitType, ua.getTuv());
+      }
+    }
```

Comment:
This is the core change of checking not just the production frontiers but also the unit type list. The other changes are just some small refactorings.